gnovm/stdlibs/std/native.go

```diff
@@ -32,6 +32,18 @@ func GetHeight(m *gno.Machine) int64 {
 	return m.Context.(ExecContext).Height
 }
 
+func GetTimestamp(m *gno.Machine) int64 {
+	return m.Context.(ExecContext).Timestamp
+}
+
+// GetFuncNameFromCallStack returns the last called function name (identifier) from the call stack.
+func GetFuncNameFromCallStack(m *gno.Machine) string {
+	if len(m.Frames) == 0 {
+		return ""
+	}
+	return string(m.Frames[0].Func.Name)
```

Comment:
Is the intention to get the name of the most recently called function that is emitting the event?  I'm not sure this will do that. Consider changing the txtar test contract to the following:
```go
func Sender(){
	SubSender()
}

func SubSender() {
	std.Emit(
		EventSender,
		[]string{
			"key1", "value1",
			"key2", "value2",
			"key3", "value3",
			"not-a-pair",		// ignore the key-value not in pairs
		},
	)
}
```
If I call `Sender` should I expect the event emitted to have a function identifier of `Sender` or `SubSender`? Currently it would be `Sender`.  It could be confusing if a function calls into various other functions, all emitting events with the same identifier.

Reply:
When I first add the function name displayed, I thought it would be enough to just show the approximate location. Therefore, I made it just print out the most recently called function name.

However, like your example, there was clearly have some confusion with this approach. So, i modified it to return the function called in the previous frame when it finds the `Emit` function while iterating through the call stack.

like this:
```go
func getPrevFunctionNameFromTarget(m *gno.Machine, targetFunc string) string {
	for i := 0; i < len(m.Frames); i++ {
		if m.Frames[i].Func.Name == gno.Name(targetFunc) && i > 0{
			return string(m.Frames[i-1].Func.Name)
		}
	}
	return ""
}

// ...
fnIdent := getPrevFunctionNameFromTarget(m, "Emit")
```

Reply:
Thanks.  You'll also need to check that the frame `Func` is not nil; it is possible for a frame to have a source other than a function call -- a `ForStmt` for example.  So emitting an event inside a loop would break this.
```go
package main

import "std"

func main() {
	for i := 0; i < 10; i++ {
		std.Emit("testing")
	}
}
```

I'm also thinking this would be better if it started iterating over the frames in the reverse order because the function frame that called `Emit` should always be near the end of the slice.  This means that it can't just get the name of `m.Frames[i-1]` once `Emit` is found. In the above loop example the i-1 frame is the for loop frame. So it needs to continue iterating until it finds a frame with a non-nil `Func`.

Reply:
@notJoon this was marked as resolved but I don't think that it is resolved. Does the code in my comment above panic when you run it?

Reply:
there's definitely a nil pointer error occurred. So, I modified the `getPrevFunctionNameFromTarget` function to include a nil check.

below is a rough form of changes
```go
func getPrevFunctionNameFromTarget(m *gno.Machine, targetFunc string) string {
    for i := len(m.Frames) - 1; i >= 0; i-- {
        if m.Frames[i].Func.Name == gno.Name(targetFunc) {
            for j := i - 1; j >= 0; j-- {
                if m.Frames[j].Func != nil {
                    return string(m.Frames[j].Func.Name)
                }
            }
        }
    }
    return ""
}
```